libraries/ESP8266WiFi/src/ESP8266WiFiSTA.cpp

```diff
@@ -467,38 +468,91 @@ IPAddress ESP8266WiFiSTAClass::dnsIP(uint8_t dns_no) {
  * @return hostname
  */
 String ESP8266WiFiSTAClass::hostname(void) {
-    return String(wifi_station_get_hostname());
+    return wifi_station_get_hostname();
 }
 
-
 /**
  * Set ESP8266 station DHCP hostname
- * @param aHostname max length:32
+ * @param aHostname max length:24
  * @return ok
  */
-bool ESP8266WiFiSTAClass::hostname(char* aHostname) {
-    if(strlen(aHostname) > 32) {
+bool ESP8266WiFiSTAClass::hostname(String aHostname) {
+  /*
+  vvvv RFC952 vvvv
+  1. A "name" (Net, Host, Gateway, or Domain name) is a text string up
+   to 24 characters drawn from the alphabet (A-Z), digits (0-9), minus
+   sign (-), and period (.).  Note that periods are only allowed when
+   they serve to delimit components of "domain style names". (See
+   RFC-921, "Domain Name System Implementation Schedule", for
+   background).  No blank or space characters are permitted as part of a
+   name. No distinction is made between upper and lower case.  The first
+   character must be an alpha character.  The last character must not be
+   a minus sign or period.  A host which serves as a GATEWAY should have
+   "-GATEWAY" or "-GW" as part of its name.  Hosts which do not serve as
+   Internet gateways should not use "-GATEWAY" and "-GW" as part of
+   their names. A host which is a TAC should have "-TAC" as the last
+   part of its host name, if it is a DoD host.  Single character names
+   or nicknames are not allowed.
+  ^^^^ RFC952 ^^^^
+
+  - 24 chars max
+  - only a..z A..Z 0..9 -]
+  - no '-' as last char
+  */
+
+    if (aHostname.length() == 0) {
+        DEBUGV("WiFi.(set)hostname(): empty name\n");
         return false;
     }
-    return wifi_station_set_hostname(aHostname);
-}
 
-/**
- * Set ESP8266 station DHCP hostname
- * @param aHostname max length:32
- * @return ok
- */
-bool ESP8266WiFiSTAClass::hostname(const char* aHostname) {
-    return hostname((char*) aHostname);
-}
+    // rework hostname to be RFC compliant
 
-/**
- * Set ESP8266 station DHCP hostname
- * @param aHostname max length:32
- * @return ok
- */
-bool ESP8266WiFiSTAClass::hostname(const String& aHostname) {
-    return hostname((char*) aHostname.c_str());
+    if (aHostname.length() > 24) {
+        // nonos-sdk allows 32, but
+        // dhcp server may require RFC compliance
+        aHostname.remove(24);
+    }
+    for (size_t i = 0; i < aHostname.length(); i++)
+        // replace unallowed chars by dashes
+        if (!isalnum(aHostname[i]) && aHostname[i] != '-')

```

Comment:
What is the reasoning behind silently changing the given hostname string?
How about pulling this first part of the code out and splitting it into two functions:
1. validateHostname() which takes a hostname string and returns true if the string complies with the RFC limitations, or false if not. Call this from WiFi.hostname(), and return false if validation fails.
2. fixHostname() which fixes a hostname string by replacing invalid chars, truncating to 24chars, etc. Make this function public. If the user wants to have a "soft" behavior, where the argument gets auto-fixed, then he can do something like WiFi.hostname(fixHostname(myhostname)).

Reply:
I too have thought to split this checking/updating out.

Users usually do not check `WiFi.hostname()` return value, on purpose, because:
- sketches would be unnecessarily complex just to take actions in case this sole function would be failing,
- there's not much to do from the sketch point of view when the API refuses this request,
- there's no easy way to verify network servers did acknowledge this request (on some dhcp servers, this hostname request is simply ignored, that's one of the reasons why they made mDNS)

In the end, I took the efficient way:
- arduino simplicity
- if a user does not know this RFC, we just stuff it to them

I'm open to any counter-argument :)

Reply:
My thinking is precisely what you said: simplicity, and the RFC, i.e.: if the hostname doesn't meet the RFC requirements, then do nothing and just fail with return false. If anyone doesn't check the return value, and/or doesn't understand why it doesn't work, we direct the person to the RFC. 
It's a bit like trying to set up a wifi connection with wpa2, and a password that is too short. 

What I'm suggesting is to just remove the code that silently replaces illegal characters with '-', and do a simple validation instead.

Reply:
> It's a bit like trying to set up a wifi connection with wpa2, and a password that is too short.

That's different, in that case, the whole sketch connectivity won't work. An `hostname(name)` failing does not prevent the sketch working and being accessible through its IP.

> if the hostname doesn't meet the RFC requirements, then do nothing and just fail with return false

I agree on that we can have silently an unexpected hostname result with this PR.

I would propose something in the middle, which is returning false when the name is not compliant (without anymore name fixing, with debug message when they are enabled), but still using this name as we were before, because dhcp servers are flexible with this RFC restriction.

Reply:
Agreed!